pkg/tracker/apitracker.go

```diff
@@ -0,0 +1,285 @@
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package tracker
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"net"
+	"net/http"
+	"strings"
+
+	"github.com/Masterminds/log-go"
+	"github.com/containers/gvisor-tap-vsock/pkg/types"
+	"github.com/docker/go-connections/nat"
+)
+
+const (
+	// The Gateway IP address that is statically reserved
+	// by DHCP and will not change. It is used to initialize
+	// the NewAPITracker.
+	GatewayBaseURL = "http://192.168.127.1:80"
+	// Tap device (eth0) IP which is also allocated to the host-switch
+	// it is statically reserved by DHCP.
+	hostSwitchIP = "192.168.127.2"
+	exposeAPI    = "/services/forwarder/expose"
+	unexposeAPI  = "/services/forwarder/unexpose"
+)
+
+var (
+	ErrAPI         = errors.New("error from API")
+	ErrExposeAPI   = fmt.Errorf("error from %s API", exposeAPI)
+	ErrUnexposeAPI = fmt.Errorf("error from %s API", unexposeAPI)
+	ErrInvalidIPv4 = errors.New("not an IPv4 address")
+)
+
+// APITracker keeps track of the port mappings and calls the
+// corresponding API endpoints that is responsible for exposing
+// and unexposing the ports on the host. This should only be used when
+// the Rancher Desktop networking is enabled and the privileged service is disabled.
+type APITracker struct {
+	baseURL     string
+	httpClient  http.Client
+	portStorage *portStorage
+	*ListenerTracker
+}
+
+// NewAPITracker creates a new instace of a API Tracker.
+func NewAPITracker(baseURL string) *APITracker {
+	return &APITracker{
+		baseURL:         baseURL,
+		httpClient:      *http.DefaultClient,
+		portStorage:     newPortStorage(),
+		ListenerTracker: NewListenerTracker(),
+	}
+}
+
+// Add a container ID and port mapping to the tracker and calls the
+// /services/forwarder/expose endpoint to forward the port mappings.
+func (a *APITracker) Add(containerID string, portMap nat.PortMap) error {
+	var errs []error
+
+	successfullyForwarded := make(nat.PortMap)
+
+	for portProto, portBindings := range portMap {
+		var tmpPortBinding []nat.PortBinding
+
+		for _, portBinding := range portBindings {
+			// The expose API only supports IPv4
+			ipv4, err := isIPv4(portBinding.HostIP)
+			if !ipv4 || err != nil {
+				continue
+			}
+
+			log.Debugf("calling %s API for the following port binding: %+v", exposeAPI, portBinding)
+
+			err = a.expose(
+				&types.ExposeRequest{
+					Local:  ipPortBuilder(portBinding.HostIP, portBinding.HostPort),
+					Remote: ipPortBuilder(hostSwitchIP, portBinding.HostPort),
+				})
+			if err != nil {
+				errs = append(errs, fmt.Errorf("exposing %+v failed: %w", portBinding, err))
+
+				continue
+			}
+
+			tmpPortBinding = append(tmpPortBinding, portBinding)
+		}
+
+		successfullyForwarded[portProto] = tmpPortBinding
+	}
+
+	a.portStorage.add(containerID, successfullyForwarded)
```

Comment:
This clobbers any potentially existing ports. Do we need to `Remove` old things first? (e.g. if it used to forward `80` and `443`, but it now gets called with `80` and `8080`. If that's not a concern, it should be documented as a requirement of `Tracker.Add`.)